akka-typed/src/main/scala/akka/typed/ScalaDSL.scala

```diff
@@ -0,0 +1,399 @@
+  }
+
+  /**
+   * This type of Behavior wraps another Behavior while allowing you to perform
+   * some action upon each received message or signal. It is most commonly used
+   * for logging or tracing what a certain Actor does.
+   */
+  case class Tap[T](f: PartialFunction[MessageOrSignal[T], Unit], behavior: Behavior[T]) extends Behavior[T] {
+    private def canonical(behv: Behavior[T]): Behavior[T] =
+      if (isUnhandled(behv)) Unhandled
+      else if (behv eq sameBehavior) Same
+      else if (isAlive(behv)) Tap(f, behv)
+      else Stopped
+    override def management(ctx: ActorContext[T], msg: Signal): Behavior[T] = {
+      f.applyOrElse(Sig(ctx, msg), unitFunction)
+      canonical(behavior.management(ctx, msg))
+    }
+    override def message(ctx: ActorContext[T], msg: T): Behavior[T] = {
+      f.applyOrElse(Msg(ctx, msg), unitFunction)
+      canonical(behavior.message(ctx, msg))
+    }
+    override def toString = s"Tap(${LN.forClass(f.getClass)},$behavior)"
+  }
+  object Tap {
+    def monitor[T](monitor: ActorRef[T], behavior: Behavior[T]): Tap[T] = Tap({ case Msg(_, msg) ⇒ monitor ! msg }, behavior)
+  }
+
+  /**
+   * This type of behavior is a variant of [[Behavior.Simple]] that does not
+   * allow the actor to change behavior. It is an efficient choice for stateless
+   * actors, possibly entering such a behavior after finishing its
+   * initialization (which may be modeled using any of the other behavior types).
+   *
+   * This behavior type is most useful for leaf actors that do not create child
+   * actors themselves.
+   */
+  case class Static[T](behavior: T ⇒ Unit) extends Behavior[T] {
+    override def management(ctx: ActorContext[T], msg: Signal): Behavior[T] = msg match {
+      case _ ⇒ Unhandled
+    }
+    override def message(ctx: ActorContext[T], msg: T): Behavior[T] = {
+      behavior(msg)
+      this
+    }
+    override def toString = s"Static(${LN.forClass(behavior.getClass)})"
+  }
+
+  /**
+   * This behavior allows sending messages to itself without going through the
+   * Actor’s mailbox. A message sent like this will be processed before the next
+   * message is taken out of the mailbox. In case of Actor failures outstanding
+   * messages that were sent to the synchronous self reference will be lost.
+   *
+   * This decorator is useful for passing messages between the left and right
+   * sides of [[ScalaDSL$.And]] and [[ScalaDSL$.Or]] combinators.
+   */
+  case class SynchronousSelf[T](f: ActorRef[T] ⇒ Behavior[T]) extends Behavior[T] {
+    private val inbox = Inbox.sync[T]("syncbox")
+    private var _behavior = f(inbox.ref)
+    private def behavior = _behavior
+    private def setBehavior(ctx: ActorContext[T], b: Behavior[T]): Unit =
+      _behavior = canonicalize(ctx, b, _behavior)
+
+    @tailrec private def run(ctx: ActorContext[T], next: Behavior[T]): Behavior[T] = {
+      setBehavior(ctx, next)
+      if (inbox.hasMessages) run(ctx, behavior.message(ctx, inbox.receiveMsg()))
```

Comment:
I found the need for this while implementing the Receiver: it speaks a query protocol on the one side and receives messages of type T on the other (erased to Any). The two halves are combined with Or in order to keep the types straight, but elements received by the RHS need to go to the LHS in order to be available for querying. Joining is not the right operation in this case.

In a hierarchical state machine I take it you can also send events between the parts, or not?

This decorator is also useful for other resequencing, like sending a message to self during a Failed signal so that common operation logic can be reused (for starting a new actor or whatever).